### PR TITLE
Add gRPC interop tests

### PR DESCRIPTION
**Pull request by JamesNK**
Addresses https://github.com/aspnet/AspNetCore/issues/11064

PR adds gRPC interop tests.

- Code is in src/Grpc directory
- Code is "disconnected" from the rest of build infrastructure through blank directory props/targets. If required the test project could be made to integrate with the rest of the build by moving the blank directory props/targets files to testassets.
- The interop client and server are completely independent projects, invoked by tests with dotnet run
- Tests run in a minute on my machine. Could speed them up by publishing client and running exe rather than calling dotnet run 18 times. I have kept things simple for now.
- Server runs with Kestrel. In the future we'll want to pass args to start server with HttpSys

Is this a good approach?
What changes need to happen elseware to run test project in ProjectTemplates job?

### CODE REVIEW

File: src/Grpc/Directory.Build.props
Comment (dougbu) [inline src/Grpc/Directory.Build.props]: Why is it necessary to completely isolate these directories (not just test/testassets) from the repo's configuration? concern here is maintenance headache of spreading the gRPC versions around. Even for test/testassets, we usually _generate_ a file containing the versions from eng/Versions.props i.e. those projects are _mostly_ isolated from the rest of the respo.

nit: If these Directory.Build.* files are truly required, use `<Project />`.

File: src/Shared/Process/ProcessEx.cs
Comment (dougbu) [inline src/Shared/Process/ProcessEx.cs]: Why is this file shared? Seems to be used in only one project.

File: src/Grpc/test/InteropTests/Infrastructure/ClientProcess.cs
Comment (dougbu) [inline src/Grpc/test/InteropTests/Infrastructure/ClientProcess.cs]: Woah, this is the start of a lot of new test infrastructure that doesn't seem to come from the gRPC repo. Don't we have anything similar already?

File: src/Grpc/test/InteropTests/InteropTestsFixture.cs
Comment (dougbu) [inline src/Grpc/test/InteropTests/InteropTestsFixture.cs]: Why isn't the fixture in the `InteropTests.Infrastructure` namespace?

File: src/Shared/Process/ProcessEx.cs
Comment (JamesNK) [reply src/Shared/Process/ProcessEx.cs]: I took it from /src/ProjectTemplates/Helpers. Will update it to use the shared file before merging.

File: src/Grpc/test/InteropTests/Infrastructure/ClientProcess.cs
Comment (JamesNK) [reply src/Grpc/test/InteropTests/Infrastructure/ClientProcess.cs]: I reuse ProcessEx. ClientProcess and WebServerProcess are help types on top of that.

File: src/Grpc/test/InteropTests/InteropTestsFixture.cs
Comment (JamesNK) [reply src/Grpc/test/InteropTests/InteropTestsFixture.cs]: Moved

File: src/Shared/Process/ProcessEx.cs
Comment (JamesNK) [reply src/Shared/Process/ProcessEx.cs]: Deleted file from templates project so they use the same shared file.

File: src/Grpc/Directory.Build.props
Comment (JamesNK) [reply src/Grpc/Directory.Build.props]: Update test project to not be completely isolated.

Client/server use shared runtime.

File: src/Grpc/test/InteropTests/Helpers/InteropTestsFixture.cs
Comment (JamesNK) [inline src/Grpc/test/InteropTests/Helpers/InteropTestsFixture.cs]: What is the best way to find the client/server projects at runtime?

Comment (JunTaoLuo) [reply src/Grpc/test/InteropTests/Helpers/InteropTestsFixture.cs]: Pass it in via an assembly attribute from MSBuild: https://github.com/aspnet/AspNetCore/blob/master/src/ProjectTemplates/test/ProjectTemplates.Tests.csproj#L89-L92

**Comment by JamesNK** [comment]
@Pilchie Some of the files in this PR are from grpc-dotnet. They all exist under `\src\Grpc\test\testassets`. None of this code runs outside of testing, and it doesn't ship.

What third-party-code actions are required?

**Comment by JamesNK** [comment]
The build passes. What are the next steps here? Is it a matter of merging the PR then updating project template job tasks to add a task to run test project? Or are there changes required to the other parts of the build to handle these new projects under /src/grpc?

I don't know what changes are required from here on. I need build SMEs to guide me 🙏

**Comment by Pilchie** [comment]
Ideally, you place the external code in a top level folder that makes it clear that it is external, and a notice added to https://github.com/aspnet/AspNetCore/blob/master/THIRD-PARTY-NOTICES.txt.

File: src/Grpc/test/InteropTests/InteropTests.cs
Comment (jkotalik) [inline src/Grpc/test/InteropTests/InteropTests.cs]: Ooo, spicy. I'll let @davidfowl get mad at you.

**Review by jkotalik — APPROVED** [review]
Don't see any reason to block this.

**Comment by JamesNK** [comment]
@JunTaoLuo Some outstanding errors. How do I resolve these.

Build error:

> C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\Bin\Roslyn\Microsoft.Managed.Core.targets(102,5): error : SourceRoot items must include at least one top-level (not nested) item when DeterministicSourcePaths is true [F:\workspace.9\_work\1\s\src\Grpc\test\testassets\InteropClient\InteropClient.csproj]
> ##[error]C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\Bin\Roslyn\Microsoft.Managed.Core.targets(102,5): error : SourceRoot items must include at least one top-level (not nested) item when DeterministicSourcePaths is true

Test error (one sample, this was thrown for for every test):

```
System.TimeoutException : The operation at /_/src/Grpc/test/InteropTests/InteropTests.cs:43 timed out after reaching the limit of 30000ms.
   at Microsoft.AspNetCore.Testing.TaskExtensions.TimeoutAfter(Task task, TimeSpan timeout, String filePath, Int32 lineNumber)
   at InteropTests.InteropTests.InteropTestCase(String name) in /_/src/Grpc/test/InteropTests/InteropTests.cs:line 45
--- End of stack trace from previous location ---
Output:
[ERROR] It was not possible to find any compatible framework version
[ERROR] The framework 'Microsoft.NETCore.App', version '5.0.0-alpha1.19521.2' was not found.
[ERROR]   - The following frameworks were found:
[ERROR]       5.0.0-alpha.1.19551.1 at [F:\workspace.9\_work\1\s\.dotnet\shared\Microsoft.NETCore.App]
[ERROR]       5.0.0-alpha.1.19562.8 at [F:\workspace.9\_work\1\s\.dotnet\shared\Microsoft.NETCore.App]
[ERROR] 
[ERROR] You can resolve the problem by installing the specified framework and/or SDK.
[ERROR] 
[ERROR] The specified framework can be found at:
[ERROR]   - https://aka.ms/dotnet-core-applaunch?framework=Microsoft.NETCore.App&framework_version=5.0.0-alpha1.19521.2&arch=x64&rid=win10-x64
```

**Comment by JunTaoLuo** [comment]
I think we might need to add something like this to the interop tests to pick up the latest shared framework: https://github.com/aspnet/AspNetCore/blob/master/src/ProjectTemplates/test/Infrastructure/TemplateTests.props.in.

Essentially we need the workaround here: https://github.com/aspnet/AspNetCore/blob/master/Directory.Build.targets#L115-L122 to be able to run on the latest shared framework. However, I believe the repo has been "isolated" so that fix isn't picked up and the latest shared framework cannot be found.

**Comment by JamesNK** [comment]
@JunTaoLuo Can you take this over - just need to get them passing in CI environment.

File: src/Grpc/test/testassets/InteropClient/InteropClient.csproj
Comment (JunTaoLuo) [inline src/Grpc/test/testassets/InteropClient/InteropClient.csproj]: 👀

File: src/Grpc/test/testassets/InteropClient/InteropClient.csproj
Comment (JamesNK) [inline src/Grpc/test/testassets/InteropClient/InteropClient.csproj]: I made the interop client and website separate from the overall build infrastructure because they needed to reference the installed ASP.NET SDK, and not ASP.NET Core projects.

It appears you're integrating them into the overall build. How did you solve the ASP.NET SDK referencing issue?

Comment (JunTaoLuo) [reply src/Grpc/test/testassets/InteropClient/InteropClient.csproj]: By ASP.NET SDK do you mean the shared framework? So far I haven't done that yet, I'm only changing how the packages are referenced.

Though I see why we need to reference the shared framework in the InteropWebsite, why is it needed for the InteropClient? It seems to me that it should really just be a regular .NET Core app.

**Comment by SteveSandersonMS** [comment]
The change to blazor.server.js looks unintentional - probably should be reverted.

**Comment by JunTaoLuo** [comment]
Yup, that's accidental. Though I feel it's really easy to unintentionally modify that file.

**Comment by JunTaoLuo** [comment]
It's now successfully building locally and on the CI. However the tests are only passing locally. On the CI it seems like the server process never ran. Also, I wonder why we didn't use the deployer pattern.

**Comment by JunTaoLuo** [comment]
The tests are up and running on the CI.

**Comment by Tratcher** [comment]
This `dotnet run` approach doesn't work in Helix, right? We shouldn't be adding any new tests at this point that don't work in Helix.

**Comment by JunTaoLuo** [comment]
It's unlikely that I'll be able to focus on this while I am working on extensions migration for the next couple of weeks. Given we have other tests in a similar position, we could leave the tests running on AzDO for now until we have a consistent story of how to test on Helix. I could potentially circle back to this when extensions work is complete.

**Comment by HaoK** [comment]
dot net run approach will work on helix, as templates test generally require installing the shared framework bits onto the helix machine and doing dotnet new/run types of commands as written.  So gRPC tests should be able to do the same thing...that is once templates tests are fully working on helix

File: src/Grpc/test/InteropTests/Helpers/WebsiteProcess.cs
Comment (HaoK) [inline src/Grpc/test/InteropTests/Helpers/WebsiteProcess.cs]: Oh, but we should revisit things like this, burning paths into the assembly doesn't work on helix, can we have this fallback to a relative path default that just works (that would be the helix path)

Comment (HaoK) [reply src/Grpc/test/InteropTests/Helpers/WebsiteProcess.cs]: Like if the attribute doesn't exist, just fallback to "/artifacts/log" which is the default place we upload logs from on helix

File: src/Grpc/test/InteropTests/InteropTests.cs
Comment (HaoK) [inline src/Grpc/test/InteropTests/InteropTests.cs]: Same thing, can we have a fallback (or better yet a config file),  @javiercn can you drive coming up with a pattern that we can use in place of this assembly metadata stuff?  Components and Templates are the main consumers of this metadata pattern I think right now

File: src/Grpc/test/InteropTests/Helpers/WebsiteProcess.cs
Comment (JunTaoLuo) [reply src/Grpc/test/InteropTests/Helpers/WebsiteProcess.cs]: The problem here isn't just about the path. Specifically, in this case, the test executable is trying to run the server/client executables but I don't think they are included in the Helix test payload. In general, we need a pattern for including other executables in test payloads.

However, in this particular case, I believe it might be possible to run the tests in process, in which case, we should be able to avoid the problem all together.

**Comment by JunTaoLuo** [comment]
@HaoK can you take a quick look at the latest commit to see if it seems sane for Helix? I think I'll merge this once the Helix tests pass on CI.

File: Directory.Build.targets
Comment (HaoK) [inline Directory.Build.targets]: nit: You call it Integration test above, and functional test here.

File: src/Shared/Process/ProcessEx.cs
Comment (HaoK) [inline src/Shared/Process/ProcessEx.cs]: merge issues in this file, does this break template tests in helix?

File: .azure/pipelines/ci.yml
Comment (HaoK) [inline .azure/pipelines/ci.yml]: If grpc tests work on helix natively, I'd recommend not adding them to the azdo test legs, and leaving this as soley for Windows_templates_tests, and you can potentially remove any of the non helix codepaths (if any) as well then.

Comment (HaoK) [reply .azure/pipelines/ci.yml]: Actually I guess you still want the non helix code paths for local dotnet test runs

Comment (JunTaoLuo) [reply .azure/pipelines/ci.yml]: There's still a need to build the shared framework before we run these interop tests which is why we put the tests here. The goal here is to future proof them so it's easy to run them on Helix as @Tratcher suggested.

File: src/Shared/Process/ProcessEx.cs
Comment (JunTaoLuo) [reply src/Shared/Process/ProcessEx.cs]: Ah yes

File: .azure/pipelines/ci.yml
Comment (HaoK) [reply .azure/pipelines/ci.yml]: So these tests are already running on helix with this PR right?  I'm just asking why do we need to add these to the non helix test job if they are passing on helix.

Comment (JunTaoLuo) [reply .azure/pipelines/ci.yml]: Fair, I'll remove this once the tests are brought up on Helix.

File: src/Grpc/test/InteropTests/Helpers/WebsiteProcess.cs
Comment (JunTaoLuo) [inline src/Grpc/test/InteropTests/Helpers/WebsiteProcess.cs]: I'll revert this commit if https://github.com/dotnet/aspnetcore/pull/20005#pullrequestreview-378116740 goes in.

**Comment by JunTaoLuo** [comment]
@HaoK The tests are passing on Helix! I'm doing one more pass to do a clean up to remove the AzDO modifications.

File: src/Grpc/test/InteropTests/Helpers/WebsiteProcess.cs
Comment (JunTaoLuo) [reply src/Grpc/test/InteropTests/Helpers/WebsiteProcess.cs]: @HaoK so eventually when #20005 goes in, it'll be correct to log directly to the current directory right?

Comment (HaoK) [reply src/Grpc/test/InteropTests/Helpers/WebsiteProcess.cs]: No, you should write to artifacts/logs, might need to create that directory

Comment (JunTaoLuo) [reply src/Grpc/test/InteropTests/Helpers/WebsiteProcess.cs]: so CurrentDirectory/artifacts/logs?

**Comment by HaoK** [comment]
Yup that should do it

**Comment by JunTaoLuo** [comment]
@JamesNK @dotnet/aspnet-build @HaoK I'm planning on merging this once tests pass. Anything else I should address before I do that?

**Comment by JamesNK** [comment]
:shipit:

I would approve but it is my own PR 😄

File: src/Grpc/test/InteropTests/Helpers/WebsiteProcess.cs
Comment (HaoK) [reply src/Grpc/test/InteropTests/Helpers/WebsiteProcess.cs]: https://github.com/dotnet/aspnetcore/pull/20005 is finally merged, so maybe try one more rebase/build to see if the logs are working if you want

File: .azure/pipelines/ci.yml
Comment (HaoK) [inline .azure/pipelines/ci.yml]: You will need to add this to all the various helix legs that you want these tests to run on (daily helix, and maybe the quarantined yml too since i imagine its possible some of these tests would get social distanced eventually too)

**Pull request merged**
